File: 0018-4sum/four-pointer.cs
public class Solution
{
    public IList<IList<int>> FourSum(int[] nums, int target)
    {
        int N = nums.Length;

        // Sort array
        Array.Sort(nums);

        // Initialize return IList
        IList<IList<int>> result = new List<IList<int>>();

        for (int i = 0; i < (N - 3); i++)
        {
            // Skip duplicates
            if ((i > 0) && (nums[i] == nums[i - 1]))
            {
                continue;
            }
            else
            {
                for (int j = (i + 1); j < (N - 2); j++)
                {
                    if ((j > (i + 1)) && (nums[j] == nums[j - 1]))
                    {
                        continue;
                    }
                    else
                    {
                        // Create two pointers
                        int k = j + 1;
                        int l = N - 1;

                        while (k < l)
                        {
                            // Calculate sum
                            long sum = (long) nums[i] + (long) nums[j] + (long) nums[k] + (long) nums[l];

                            if (sum < (long) target)
                            {
                                k++;
                            }
                            else if (sum > (long) target)
                            {
                                l--;
                            }
                            else
                            {
                                // Add quadtriplet
                                result.Add(new List<int> {nums[i], nums[j], nums[k], nums[l]});

                                k++;
                                l--;

                                // Skip duplicates
                                while ((k < l) && (nums[k] == nums[k - 1]))
                                {
                                    k++;
                                }

                                //Skip duplicates for 4th number
                                while ((k < l) && (nums[l] == nums[l + 1]))
                                {
                                    l--;
                                }

                            }
                        }
                    }
                }
            }
        }

        return result;
    }
}

File: 0018-4sum/four-pointer.rs
impl Solution
{
    pub fn four_sum(nums: Vec<i32>, target: i32) -> Vec<Vec<i32>>
    {
        let n: usize = nums.len() as usize;

        // Sort array
        let mut arr: Vec<i32> = nums;
        arr.sort();

        // Initialize return vector
        let mut result: Vec<Vec<i32>> = Vec::new();

        if (n >= 4)
        {
            for i in 0..(n-3)
            {
                // Skip duplicates
                if ((i > 0) && (arr[i] == arr[i - 1]))
                {
                    continue;
                }
                else
                {
                    for j in (i+1)..(n-2)
                    {
                        if ((j > (i + 1)) && (arr[j] == arr[j - 1]))
                        {
                            continue;
                        }
                        else
                        {
                            // Create two pointers
                            let mut k: usize = (j + 1);
                            let mut l: usize = (n - 1);

                            while (k < l)
                            {
                                // Calculate sum
                                let sum: i64 = (arr[i] as i64) + (arr[j] as i64) + (arr[k] as i64) + (arr[l] as i64);

                                if (sum < (target as i64))
                                {
                                    k += 1;
                                }
                                else if (sum > (target as i64))
                                {
                                    l -= 1;                                
                                }
                                else
                                {
                                    // Add quadtriplet
                                    result.push(vec![arr[i], arr[j], arr[k], arr[l]]);

                                    k += 1;
                                    l -= 1;

                                    // Skip duplicates
                                    while ((k < l) && (arr[k] == arr[k - 1]))
                                    {
                                        k += 1;
                                    }

                                    //Skip duplicates for 4th number
                                    while ((k < l) && (arr[l] == arr[l + 1]))
                                    {
                                        l -= 1;
                                    }
                                }
                            }
                        }
                    }
                }
            }   
        }

        return result;
    }
}